### PR TITLE
Make ApplyBucketsWork independent of merges in progress

### LINKED ISSUE #3044: Allow ApplyBucketsWork to use local HAS
Right now ApplyBucketsWork assumes it receives the HAS file from the history archives, so it performs `containsValidBuckets` check. The problem is that in a few places in the code we actually use local HAS files which are valid and trusted, but might not meet the archive validity criteria. So we should move `containsValidBuckets` check outside of the ApplyBucketsWork, and into catchup instead.

### PR DESCRIPTION
**Pull request by marta-lokhova**
Resolves #3044

This change removes the dependency on future buckets state from ApplyBucketsWork. Previously, ApplyBucketsWork would expect the future buckets in HAS to be either resolved or clear, as it expected HAS from an untrusted archive. This has changed, since we now use the local HAS in ApplyBucketsWork as well (in captive core startup, for example), which may have merges in progress. So this change moves the HAS validation into DownloadBucketsWork instead.

Note that it also fixes the race condition that impacts older (before protocol 12) buckets: the HAS validation expects some merges to be resolved, however `prepareForPublish` restarts the merges in progress, but does not necessarily wait for them to finish.

### CODE REVIEW

File: src/historywork/DownloadBucketsWork.cpp [hunk 1]
```diff
@@ -17,8 +17,8 @@ namespace stellar
 DownloadBucketsWork::DownloadBucketsWork(
     Application& app, std::map<std::string, std::shared_ptr<Bucket>>& buckets,
     std::vector<std::string> hashes, TmpDir const& downloadDir,
-    std::shared_ptr<HistoryArchive> archive)
-    : BatchWork{app, "download-verify-buckets"}
+    std::shared_ptr<HistoryArchive> archive, std::function<bool()> verifyCb)
```
Comment (MonsieurNicolas) [inline src/historywork/DownloadBucketsWork.cpp:20]: I think it's time to actually do https://github.com/stellar/stellar-core/pull/3030#discussion_r628649644 what do you think?

Comment (marta-lokhova) [reply src/historywork/DownloadBucketsWork.cpp]: I added a new small work class `WorkWithCallback` to address this, let me know what you think.

File: src/work/WorkWithCallback.cpp [hunk 2]
```diff
@@ -0,0 +1,48 @@
+// Copyright 2021 Stellar Development Foundation and contributors. Licensed
```
Comment (MonsieurNicolas) [inline src/work/WorkWithCallback.cpp:1]: need to update Visual Studio files

**Review by MonsieurNicolas — CHANGES_REQUESTED** [review]
this looks good. Just forgot about VC++ :)

File: src/work/WorkWithCallback.cpp (hunk 2 above)
Comment (marta-lokhova) [reply src/work/WorkWithCallback.cpp:1]: Fixed.

**Comment by MonsieurNicolas** [comment]
r+ 55e5ddd60eed626ad23d1d3784d68c3f9572c618

**Pull request merged**